pkg/api/pool.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// pool.go contains an implementation of a pool of workers that execute aspects.
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"istio.io/mixer/pkg/adapterManager"
+	"istio.io/mixer/pkg/aspect"
+	"istio.io/mixer/pkg/attribute"
+	"istio.io/mixer/pkg/expr"
+)
+
+// result holds the values returned by the execution of an adapter on a thread in the pool
+type result struct {
+	name string
+	out  *aspect.Output
+	err  error
+}
+
+// enqueueFunc is a handle into the pool and is used to add work to the pool. It will block until a worker in the pool
+// is available to execute the work item.
+type enqueueFunc func(ctx context.Context, config *aspect.CombinedConfig)
+
+// task describes one unit of work to be executed by the pool
+type task struct {
+	ctx    context.Context
+	mngr   *adapterManager.Manager
+	cfg    *aspect.CombinedConfig
+	ab     attribute.Bag
+	mapper expr.Evaluator
+	done   chan<- result
+}
+
+// pool refers to a group of goroutinues that execute tasks and return results
+type pool struct {
+	work chan<- task     // Used to hand work items to the workers in the pool
+	quit chan<- struct{} // Used to kill the workers in the queue
+	wg   *sync.WaitGroup // Used to block shutdown until all workers complete
```

Comment:
[`Waitgroup`s can't be copied after use](https://golang.org/pkg/sync/#WaitGroup), but we're handing it off to other goroutines, so it must be a pointer, right? Or, at least it must be a pointer when it's handed to the goroutine, the pool itself can hold the instance directly since the pool is only passed around as a pointer.